### ISSUE TITLE
Rule proposal: lint readonly not-assigned properties

### ISSUE THREAD

**Issue by jendrikw** [sindresorhus/eslint-plugin-unicorn#2954]
### Description

I have a readonly class property that I forgot to assign in the constructor. Even with `strictPropertyInitialization`, typescript does not warn when the type of the property allows undefined.

The result is that the property can only ever be undefined, which is probably not what the author intended (and in such a case should be typed as just `undefined`.)

This would depend on typescript for the `readonly` syntax.

(Rejected by typescript-eslint https://github.com/typescript-eslint/typescript-eslint/issues/12315)

### Examples

```ts
// ❌
class A {
  readonly x: string | undefined;
  constructor(x: string | undefined) {
    // complex logic
    if (x == "foo") {
      x = "bar"
    }
    // forget to intialize x
  }
}

// ✅
class A {
  readonly x: string | undefined;
  constructor(x: string | undefined) {
    // complex logic
    if (x == "foo") {
      x = "bar"
    }
    this.x = x;
  }
}
```

### Proposed rule name

no-uninitialized-readonly

### Additional Info

Rough sketch, needs testing:

```js
        /**
         * @param {import("eslint").Rule.RuleContext} context
         */
        create(context) {
            /**
             * @type {Array<{readonlyFields: Map<string, {initialized: boolean, node: import("@typescript-eslint/utils").TSESTree.PropertyDefinition}>, node: import("@typescript-eslint/utils").TSESTree.ClassBody}>}
             */
            const classStack = [];

            // noinspection JSUnusedGlobalSymbols
            return {
                /**
                 * @param {import("estree").AssignmentExpression} node
                 */
                AssignmentExpression(node) {
                    const currentClass = classStack.at(-1);

                    if (!currentClass) {
                        return;
                    }
                    const ancestors = context.sourceCode.getAncestors(node);

                    const inConstructor = ancestors.some((a) => a.type === "MethodDefinition" && a.kind === "constructor");

                    if (!inConstructor) {
                        return;
                    }

                    // matches:
                    // this.x = ...
                    if (
                        node.left.type === "MemberExpression" &&
                        !node.left.computed &&
                        node.left.object.type === "ThisExpression" &&
                        node.left.property.type === "Identifier"
                    ) {
                        const name = node.left.property.name;

                        const field = currentClass.readonlyFields.get(name);

                        if (field) {
                            field.initialized = true;
                        }
                    }
                },

                /**
                 * @param {import("@typescript-eslint/utils").TSESTree.ClassBody} node
                 */
                ClassBody(node) {
                    /** @type {Map<string, {initialized: boolean, node: import("@typescript-eslint/utils").TSESTree.PropertyDefinition}>} */
                    const readonlyFields = new Map();

                    // collect readonly instance fields
                    for (const element of node.body) {
                        if (element.type === AST_NODE_TYPES.PropertyDefinition && !element.static && element.readonly && element.key.type === AST_NODE_TYPES.Identifier) {
                            readonlyFields.set(element.key.name, {
                                initialized: !!element.value,
                                node: element,
                            });
                        }
                    }

                    classStack.push({
                        node,
                        readonlyFields,
                    });
                },

                "ClassBody:exit"() {
                    const currentClass = classStack.pop();
                    // report uninitialized fields
                    for (const [name, info] of currentClass?.readonlyFields.entries() ?? []) {
                        if (!info.initialized) {
                            context.report({
                                data: {name},
                                messageId: "uninitialized",
                                node: info.node,
                            });
                        }
                    }
                }
            };
        }
```

**Comment by fregante** [comment]
I'm kinda surprised that TS doesn't catch this.

https://www.typescriptlang.org/play/?#code/FAYwNghgzlAECCsDexawE4FMIBMD2AdmAJ6wAeAXLFAC7oCWBA5rAD6wCuBOmAZo5hwBuVLBCFa6DiBp50ACkrU6jFuy49+BQQEpkotAHpDYvAFsADmExlYYPE3ogDser1iLYAXi+wARLx4eH56KGjh5N7+AEYQ6H4uAL4uxrCB6EyYNLCyrgQ09BBg9ABemOSiyYlAA

However it doesn't report this either 🤷‍♂️ 

```ts
function foo() {
	let x: true | undefined;
	return x;
}
```

**Comment by sindresorhus** [comment]
Thanks, but declining.

This is a type-level problem, and this plugin is deliberately pure-AST: no rule uses the TypeScript type checker. The only useful target is "type includes `undefined`", and detecting that correctly (e.g. `type Maybe<T> = T | undefined`) needs the checker, so we could only ship a leaky textual approximation that also requires a pile of guards (`?`, `!`, `declare`/abstract) to avoid false positives.

The right home is typescript-eslint, which sadly already declined it (#12315). It's worth re-raising there, since with the checker it'd be correct. And since TS misses the analogous local-variable case too, it's really a diagnostics gap worth filing upstream on TypeScript itself.
